pkg/inventory/inventory-client.go

```diff
@@ -24,455 +22,287 @@ import (
+			ui.ClusterObj = obj
+		}
+		uObj, err := cic.toUnstructured(ui)
+		if err != nil {
+			return err
+		}
+		var newObj *unstructured.Unstructured
+		if create {
+			newObj, err = cic.client.Namespace(uObj.GetNamespace()).Create(ctx, uObj, metav1.CreateOptions{})
+			if err != nil {
+				return err
+			}
+		} else {
+			newObj, err = cic.client.Namespace(uObj.GetNamespace()).Update(ctx, uObj, metav1.UpdateOptions{})
+			if err != nil {
+				return err
+			}
+		}
+		ui.ClusterObj = newObj
+		return nil
+	})
 }
 
-func (cic *ClusterClient) ListClusterInventoryObjs(ctx context.Context) (map[string]object.ObjMetadataSet, error) {
-	// Define the mapping
-	mapping, err := cic.mapper.RESTMapping(cic.gvk.GroupKind(), cic.gvk.Version)
-	if err != nil {
-		return nil, err
+// UpdateStatus updates the status of the in-cluster inventory
+// Performs a simple in-place update on the unstructured object
+func (cic *UnstructuredClient) UpdateStatus(ctx context.Context, inv Inventory, _ UpdateOptions) error {
+	ui, ok := inv.(*UnstructuredInventory)
+	if !ok {
+		return fmt.Errorf("expected UnstructuredInventory")
 	}
-
-	// retrieve the list from the cluster
-	clusterInvs, err := cic.dc.Resource(mapping.Resource).List(ctx, metav1.ListOptions{})
-	if err != nil && !apierrors.IsNotFound(err) {
-		return nil, err
+	if ui == nil {
+		return fmt.Errorf("inventory is nil")
 	}
-	if apierrors.IsNotFound(err) {
-		return map[string]object.ObjMetadataSet{}, nil
-	}
-
-	identifiers := make(map[string]object.ObjMetadataSet)
-
-	for i, inv := range clusterInvs.Items {
-		invName := inv.GetName()
-		identifiers[invName] = object.ObjMetadataSet{}
-		invObj, err := cic.InventoryFactoryFunc(&clusterInvs.Items[i])
+	return retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		obj, err := cic.client.Namespace(ui.Namespace()).Get(ctx, ui.Name(), metav1.GetOptions{})
 		if err != nil {
-			return nil, err
+			return err
 		}
-		wrappedInvObjSlice, err := invObj.Load()
+		ui.ClusterObj = obj
+		uObj, err := cic.toUnstructured(ui)
 		if err != nil {
-			return nil, err
+			return err
 		}
-		identifiers[invName] = append(identifiers[invName], wrappedInvObjSlice...)
-	}
-
-	return identifiers, nil
-}
-
-// deleteInventoryObjByName deletes the passed inventory object from the APIServer, or
-// an error if one occurs.
-func (cic *ClusterClient) deleteInventoryObjByName(ctx context.Context, obj *unstructured.Unstructured, dryRun common.DryRunStrategy) error {
-	if obj == nil {
-		return fmt.Errorf("attempting delete a nil inventory object")
-	}
-	if dryRun.ClientOrServerDryRun() {
-		klog.V(4).Infof("dry-run delete inventory object: not deleted")
+		// Update observedGeneration, if it exists
+		_, ok, err = unstructured.NestedInt64(uObj.Object, "status", "observedGeneration")
+		if err != nil {
+			return err
+		}
+		if ok {
+			err = unstructured.SetNestedField(uObj.Object, uObj.GetGeneration(), "status", "observedGeneration")
```

Comment:
nit: since you're already calling NestedInt64 up there, you might as well just use the returned value, rather than making uObj.GetGeneration() do it again.

Reply:
Ah. I misread the code.

Different question then: 
Why do we care if the status.observedGeneration already exists?
Is this to prevent updating the field on objects that doesn't have it in the scheme, like the ConfigMap?
If so, this seems like a sort of round-about way of doing that. I would sorta expect this to be avoided by checking the status policy, and requiring the status policy be None for ConfigMap inventories.

Reply:
Fair enough. This has the gotcha that the resource needs to have a default value in the CRD, otherwise it will never get updated. But that's probably fine for now.

Is it possible that this is a breaking behavioral change for the ConfigMap inventory? I don't know if we had any tests for ConfigMap + StatusPolicyAlll.

Reply:
Deduping this with the other thread about inventory status.